Add TextContent.Annotations

stephentoub opened the pull request:
Adds support to M.E.AI.Abstractions for annotations.

Some design notes:
- Right now, I have the Annotations property on TextContent. We can push it down to the base later without it being a breaking change.
- Most services only support citations. OpenAI has a broader notion of annotations, and it looks like Anthropic will be adding other kinds of annotations as well. I modeled this by having a base AIAnnotation and a derived CitationAnnotation.
- Different services are very varied on what they support with citations. I went back and forth on whether to have CitationAnnotation just be a flat set of properties, or to try to model it as further derived types, e.g. UrlCitationAnnotation. I've gone with the former, but could be swayed to the latter if y'all think that's more wise. Note that there's a variety of things OpenAI doesn't support or expose that others do, e.g. Google, Anthropic, and AWS all support including snippets of the cited content; I've included such things here, as well.

SteveSandersonMS commented on src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/CitationSourceLocation.cs:
```diff
@@ -0,0 +1,26 @@
+    public int? PageEnd { get; set; }
+
+    /// <summary>Gets or sets the block or chunk at which the cited material begins.</summary>
+    public int? BlockStart { get; set; }
```
Are these relative to the page or absolute?

SteveSandersonMS commented on src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/CitationSourceLocation.cs:
```diff
@@ -0,0 +1,26 @@
+    public int? BlockEnd { get; set; }
+
+    /// <summary>Gets or sets the character at which the cited material begins.</summary>
+    public int? CharacterStart { get; set; }
```
Are these relative to the block or absolute?

SteveSandersonMS commented on src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/CitationSourceLocation.cs:
```diff
@@ -0,0 +1,26 @@
+public class CitationSourceLocation
+{
+    /// <summary>Gets or sets the page at which the cited material begins.</summary>
+    public int? PageStart { get; set; }
```
Are these zero or one based?

SteveSandersonMS commented on src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/CitationSourceLocation.cs:
```diff
@@ -0,0 +1,26 @@
+
+    /// <summary>Gets or sets the character at which the cited material ends.</summary>
+    public int? CharacterEnd { get; set; }
+}
```
This is a tricky area. I know that if this is the data the underlying providers provide, we have to expose it in the same shape. But I feel for the poor app devs who have to somehow make a UI function on top of it - that's a lot of different combinations of data that might/might not be provided, and hard to interpret that info (e.g., what actually is a block, will the set of provided numbers change over time, etc.)

Do you think there will be enough consistency across providers to make this useful? If not I guess another option is to rely on people using `AIAnnotation.RawRepresentation` to get more provider-specific info.

SteveSandersonMS commented on src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/CitationSourceLocation.cs:
```diff
@@ -0,0 +1,26 @@
+    public int? BlockStart { get; set; }
+
+    /// <summary>Gets or sets the block or chunk at which the cited material ends.</summary>
+    public int? BlockEnd { get; set; }
```
Inclusive or exclusive?

SteveSandersonMS commented on src/Libraries/Microsoft.Extensions.AI.OpenAI/OpenAIAssistantChatClient.cs:
```diff
@@ -203,18 +203,53 @@ public async IAsyncEnumerable<ChatResponseUpdate> GetStreamingResponseAsync(
+                        if (!string.IsNullOrWhiteSpace(tau.InputFileId))
+                        {
+                            fileId = tau.InputFileId;
+                            toolName = "file_search";
```
Just curious, what do you expect developers to do with this tool name information? If we chose not to set it in these cases, would there be any drawback?

Asking because it feels like annotations could have file IDs even if they came from other tools, but maybe that's a problem for the future and not for now.

stephentoub replied:
Just to be able to distinguish the source of the file ID, if it matters. InputFileId is documented to come from file_search.

SteveSandersonMS reviewed (approved):
Looks good!

I added some questions/comments about bits where I think the meaning may be hard to interpret but the broad concept makes sense and feels valuable.

> Note that there's a variety of things OpenAI doesn't support or expose that others do, e.g. Google, Anthropic, and AWS all support including snippets of the cited content; I've included such things here, as well.

That sounds wise to me. Including a "snippet" is extremely helpful to the app developer - maybe OpenAI (or its client library) will add that later. It's good we're not restricted to a pure lowest-common-denominator set of information and instead are modelling what's common across a broad range of providers.

stephentoub commented on src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/AIAnnotation.cs:
```diff
@@ -0,0 +1,55 @@
+
+    /// <summary>
+    /// Gets or sets an optional placeholder or marker text that was present in the original content
+    /// to indicate the citation (e.g., "〖4:0+source.json〗" or "[1]").
```
I think I'll just remove the placeholder. It can be added later if necessary, but I believe it's only relevant for assistants, not for chat completion nor responses, and I've not come across anyone else that uses this mechanism.

rogerbarreto commented:
Worth checking the [W3C Annotation Model](https://www.w3.org/TR/annotation-model/) 

Could be helpful on defining the specialized annotations which in this document are named like `Selectors`

- TextPositionSelector
- SvgSelector
- FragmentSelector
- CssSelector
- XPathSelector
- RangeSelector ...

stephentoub commented on src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/AIAnnotation.cs:
@SteveSandersonMS, any opinion on this?

stephentoub commented:
I've addressed the feedback.

@rogerbarreto and @SteveSandersonMS, could you both please take another look? Including at my naming choices as part of addressing Roger's feedback.

@eiriktsarpalis, can you also review please?

Thanks.

SteveSandersonMS commented:
@stephentoub Looks good to me. I see you've ended up going for the most general design. Given that it's a relatively advanced feature, that seems a reasonable tradeoff even if people have to do a bit of casting to access character indices. Moving annotations to `AIContent` sounds good too if you think that's where it would end up going in the long run anyway.

eiriktsarpalis commented on src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/TextSpanAnnotatedRegion.cs:
```diff
@@ -0,0 +1,32 @@
+/// <summary>Describes a location in the associated <see cref="AIContent"/> based on starting and ending character indices.</summary>
+/// <remarks>This <see cref="AnnotatedRegion"/> typically applies to <see cref="TextContent"/>.</remarks>
+[DebuggerDisplay("[{StartIndex}, {EndIndex})")]
+public sealed class TextSpanAnnotatedRegion : AnnotatedRegion
```
Nvm, just saw that this is part of a hierarchy.

stephentoub commented on src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/AIAnnotation.cs:
```diff
@@ -0,0 +1,42 @@
+    /// The most common form of <see cref="AnnotatedRegion"/> is <see cref="TextSpanAnnotatedRegion"/>, which provides starting and ending character indices
+    /// for <see cref="TextContent"/>.
+    /// </remarks>
+    public AnnotatedRegion? AnnotatedRegion { get; set; }
```
> I believe we might cases where we have multiple selections for the same annotation, 80% > will be single, which I think we can have a helper property like we do for other abstractions.

I'm worried we're continuing to make the API more and more complicated to handle future possibilities.

Are there APIs you're thinking of that do this?

> Maybe Span or Selection instead of Region ?

Here's o3's analysis:

---
<html>
<body>
<!--StartFragment--><html><head></head><body><p><strong>Short answer:</strong> use <strong><code inline="">region</code></strong>.</p>
<hr>
<h3>Why “region” scales cleanly across media types</h3>

Candidate | Common associations | Where it feels awkward | Hidden baggage
-- | -- | -- | --
span | Text offsets (“character span”, NLP “entity span”) and sometimes “temporal span” | Spatial areas in images/video (“a bounding-box span” sounds odd) | Overlaps with the HTML <span> element and with tracing/span concepts in distributed-tracing APIs
selection | Any user-highlighted area, clipboard APIs | Objects that were never user-selected (e.g., server-side auto-generated citations) | Connotes something ephemeral and UI-driven rather than a persisted addressable range
region | “Region of interest” (images/CV), “audio region” (DAWs), “time region” (video editors), “text region” (editors) | Rarely—“region of text” is still intelligible | Minimal; seldom used as a reserved keyword in popular web specs


<p>Because <strong>region</strong> already means “a contiguous, addressable area inside larger content” in graphics, audio, and video tool-chains, it maps naturally to the non-text modalities you care about. Meanwhile, developers will still parse “region.start_char … region.end_char” for text without surprise.</p>
<h3>Additional tips</h3>
<ol>
<li>
<p><strong>Define modality-specific coordinate schemes.</strong></p>
<pre><code class="language-jsonc">// Example: one shape with multiple coordinate systems
{
  "region": {
    "text": { "startChar": 120, "endChar": 145 },
    "image": { "x": 0.25, "y": 0.40, "width": 0.15, "height": 0.10 },
    "video": { "startTime": 12.5, "endTime": 13.8 }
  }
}
</code></pre>
</li>
<li>
<p><strong>Keep the noun stable, vary the qualifier.</strong></p>
<ul>
<li>
<p><code inline="">region.coordinates</code>, <code inline="">region.mimeType</code>, <code inline="">region.confidence</code>, etc.<br>
This preserves discoverability in IDE auto-completion.</p>
</li>
</ul>
</li>
<li>
<p><strong>Document plural behavior.</strong><br>
If an annotation can apply to multiple disjoint areas, prefer an array (<code inline="">regions: [...]</code>) rather than overloading one word.</p>
</li>
<li>
<p><strong>Consider a helper alias.</strong><br>
If many text-only clients exist, you can expose <code inline="">span</code> as a <em>convenience</em> field that simply mirrors <code inline="">region.text</code>, without letting it leak into core concepts.</p>
</li>
</ol>
<hr>
<p>By choosing <strong>region</strong> you minimize mental context-switching when developers move between text, image, audio, and video workflows, and you avoid colliding with well-known UI or HTML semantics.</p></body></html><!--EndFragment-->
</body>
</html>

rogerbarreto replied:
Region it is 😊

rogerbarreto replied:
@stephentoub Actually after looking better into the [W3C Annotation selectors](https://www.w3.org/TR/annotation-model/#text-quote-selector), I'm more convinced that we should consider something more like the `Selectors` approach.

For example, a `AnnotatedTextPosition` is a unique and fixed place in the target document while an `AnnotatedQuote` or `AnnotatedFragment` which will one but would actually by its nature be parsed as multiple places in the same document.

stephentoub replied:
> I'm more convinced that we should consider something more like the Selectors approach.

Meaning:
```C#
public IList<AnnotatedRegion>? AnnotatedRegions { get; set; }
```
?

rogerbarreto replied:
> Meaning

Yes, I think this make's sense as we can have multiple specialized AnnotatedRegion, to start with the `TextSpanAnnotatedRegion` as the proposal here.
